### PR TITLE
feat(tracing): route/service scoping

### PR DESCRIPTION
**Pull request by StarlightIbuki**
implement KAG-244

### CODE REVIEW

File: kong/tracing/tracer.lua [hunk 1]
```diff
@@ -0,0 +1,97 @@
+local _M = {}
```
Comment (mayocream) [inline kong/tracing/tracer.lua:1]: Is it possible to combine the `tracer.lua` with `pdk/tracing.lua`?

Comment (StarlightIbuki) [reply kong/tracing/tracer.lua]: done.

File: spec/03-plugins/37-opentelemetry/04-exporter_spec.lua [hunk 2]
```diff
@@ -52,17 +52,22 @@ for _, strategy in helpers.each_strategy() do
...
+                                              paths = { "/" }}))
 
-      bp.plugins:insert({
+      assert(bp.routes:insert({ service = http_srv,
+                                protocols = { "http" },
+                                paths = { "/no_plugin" }}))
+
+      assert(bp.plugins:insert({
         name = "opentelemetry",
+        route = route,
```
Comment (mayocream) [inline spec/03-plugins/37-opentelemetry/04-exporter_spec.lua:65]: Do not apply otel plugin on the route by default, add an arg in `setup_instrumentations` to configure it instead.

File: spec/03-plugins/37-opentelemetry/04-exporter_spec.lua [hunk 3]
```diff
@@ -140,6 +145,52 @@ for _, strategy in helpers.each_strategy() do
       end)
     end)
 
+    describe("#scoping", function ()
```
Comment (mayocream) [inline spec/03-plugins/37-opentelemetry/04-exporter_spec.lua:148]: Otel plugin applied on services also need to be tested

File: spec/03-plugins/37-opentelemetry/04-exporter_spec.lua [hunk 4]
```diff
@@ -140,6 +145,52 @@ for _, strategy in helpers.each_strategy() do
...
+          path    = "/no_exist",
+        })
+        assert.res_status(404, r)
+
+        -- close client connection
+        cli:close()
+
+        local ok, err = thread:join()
+
+        ngx.sleep(1000)
```
Comment (mayocream) [inline spec/03-plugins/37-opentelemetry/04-exporter_spec.lua:187]: use wait_util

Comment (StarlightIbuki) [reply spec/03-plugins/37-opentelemetry/04-exporter_spec.lua:187]: This is a mistakenly committed debugging code.

**Comment by fffonion** [comment]
I feel like we are mixing the source of truth of "whether to trace or not" in different places. Previously when
otel plugin can only be global:
1. we have a `opentelemetry_tracing` config that tells Kong whether to collect specific instrumentations
2. we have otel plugin that controls whether and where a collected spans needs to be flushed to

In this PR we have an additional source of truth on "whether to trace or not" depending on if any of a tracing
plugin (otel in this case) is enabled on current path, and that conflicts with fact that it's previously only
controlled by `opentelemetry_tracing`. So i would suggest we refactor the implementation a bit so that:
1. we still rely on `opentelemetry_tracing` config that tells Kong whether to collect specific instrumentations
2. rely on plugins_iterator to pick a plugin on current request, if such otel plugin exists, then its `log` phase
phase handler is executed, and spans are flushed.

so that we don't need complex logic over iterration over plugins and cache invalidations.

**Review by mayocream — APPROVED** [review]

**Pull request merged**